common/quotas/clocked_rate_limiter_test.go

```diff
@@ -151,3 +151,81 @@ func TestClockedRateLimiter_Wait_DeadlineWouldExceed(t *testing.T) {
 	t.Cleanup(cancel)
 	assert.ErrorIs(t, rl.Wait(ctx), quotas.ErrRateLimiterReservationWouldExceedContextDeadline)
 }
+
+// test that reservations for 1 token ARE unblocked by RecycleToken
+func TestClockedRateLimiter_Wait_Recycle(t *testing.T) {
+	t.Parallel()
+	ts := clock.NewEventTimeSource()
+	rl := quotas.NewClockedRateLimiter(rate.NewLimiter(1, 1), ts)
+	ctx := context.Background()
+
+	// take first token
+	assert.NoError(t, rl.Wait(ctx))
+
+	// wait for next token and report when success
+	waiting := make(chan struct{})
+	done := make(chan struct{})
+	go func() {
+		go func() {
+			assert.NoError(t, rl.Wait(ctx))
+			done <- struct{}{}
+		}()
+
+		// The waiting channel isn't enough by itself to confirm waiting, because
+		// the inner go func can be called without the rl.Wait() having finished.
+		// Use time.Sleep to allow rl.Wait() to complete before signalling `waiting`.
+		// 1-millisecond sleep succeeded >50,000x, whereas 1-nanosecond sleep deadlocks.
+		time.Sleep(time.Millisecond)
```

Comment:
I would use 10ms as a value for "wait for another goroutine to start and get blocked in something" just to be absolutely sure. remember flakiness in CI is going to be a lot worse than locally